Data/IP/Addr.hs

```diff
@@ -258,85 +259,186 @@ showIPv6 ip@(IP6 (a1,a2,a3,a4))
-    toWord32 _             = error "toWord32"
+    !w = (fromIntegral a1 .&. 0xff) `unsafeShiftL` 24 .|.
+         (fromIntegral a2 .&. 0xff) `unsafeShiftL` 16 .|.
+         (fromIntegral a3 .&. 0xff) `unsafeShiftL`  8 .|.
+         (fromIntegral a4 .&. 0xff)
+toIPv4 _ = error "IPv4 field list length != 4"
+{-# INLINE toIPv4 #-}
 
 {-|
-  The 'toIPv6' function takes a list of 'Int' and returns 'IPv6'.
+  The 'toIPv4w' function constructs the 'IPv4' address corresponding to the
+  given 'Word32' value.  Unlike the 'fromHostAddress' function, it is strict in
+  the input value, which here is in host byte order.
+
+>>> toIPv4w 0xc0000201
+192.0.2.1
+
+@since 1.7.9
+-}
+toIPv4w :: Word32 -> IPv4
+toIPv4w !w = IP4 w
+{-# INLINE toIPv4w #-}
+
+{-|
+  The 'toIPv6' function returns the 'IPv6' address corresponding to the given
+  list of eight 16-bit 'Int's.  The function is strict in the eight elements of
+  the list.  An error is returned if the list has a differnet length.  The
+  input elements are in host byte order and are silently truncated to their 16
+  least-signicant bits before they are combined to form the IPv6 address.
 
 >>> toIPv6 [0x2001,0xDB8,0,0,0,0,0,1]
 2001:db8::1
 -}
 toIPv6 :: [Int] -> IPv6
-toIPv6 ad = IP6 (x1,x2,x3,x4)
+toIPv6 [i1,i2,i3,i4,i5,i6,i7,i8] = IP6 (x1,x2,x3,x4)
   where
-    [x1,x2,x3,x4] = map toWord32 $ split2 ad
-    split2 [] = []
-    split2 x  = take 2 x : split2 (drop 2 x)
-    toWord32 [a1,a2] = fromIntegral $ shift a1 16 + a2
-    toWord32 _       = error "toWord32"
+    !x1 = fromIntegral $ (i1 .&. 0xffff) `unsafeShiftL` 16 .|. (i2 .&. 0xffff)
+    !x2 = fromIntegral $ (i3 .&. 0xffff) `unsafeShiftL` 16 .|. (i4 .&. 0xffff)
+    !x3 = fromIntegral $ (i5 .&. 0xffff) `unsafeShiftL` 16 .|. (i6 .&. 0xffff)
+    !x4 = fromIntegral $ (i7 .&. 0xffff) `unsafeShiftL` 16 .|. (i8 .&. 0xffff)
+toIPv6 _ = error "toIPv6 field list length != 8"
+{-# INLINE toIPv6 #-}
 
 {-|
-  The 'toIPv6b' function takes a list of 'Int'
-  where each member repserents a single byte and returns 'IPv6'.
+  The 'toIPv6b' function returns the IPv6 address corresponding to the given
+  list of sixteen 'Int' octets.  The function is strict in the sixteen elements
+  of the list.  An error is returned if the list has a differnet length.  The
+  input elements are silently truncated to their 8 least-signicant bits before
+  they are combined to form the IPv6 address.
 
 >>> toIPv6b [0x20,0x01,0xD,0xB8,0,0,0,0,0,0,0,0,0,0,0,1]
 2001:db8::1
 -}
 toIPv6b :: [Int] -> IPv6
-toIPv6b ad = IP6 (x1,x2,x3,x4)
+toIPv6b [ h11, h12, l11, l12, h21, h22, l21, l22
+        , h31, h32, l31, l32, h41, h42, l41, l42 ] = IP6 (x1,x2,x3,x4)
   where
-    [x1,x2,x3,x4] = map toWord32 $ split4 ad
-    split4 [] = []
-    split4 x  = take 4 x : split4 (drop 4 x)
-    toWord32 [a1,a2,a3,a4] = fromIntegral $ shift a1 24 + shift a2 16 + shift a3 8 + a4
-    toWord32 _       = error "toWord32"
+    !x1 = fromIntegral $ (h11 .&. 0xff) `unsafeShiftL` 24 .|.
```

Comment:
ditto

Reply:
Again, these are in fact needed.

Reply:
For example (ghci):
```
λ> :set -XBangPatterns                       
λ> let !x = (undefined, undefined, undefined, undefined)
λ> let y = seq x 1
λ> y
```

Reply:
Or are you assuming that you're also enabling `Strict` by default???

Reply:
You've merged that change I guess, so if it is staying, then yes, we don't need `!` in let bindings, they are there by default.

But I see that the 7.x builds are generating spurious warnings from "cabal check", despite the "impl(ghc) >= 8" guard. Is there some way to avoid that, and have all the builds be green?

Reply:
If you want to do this kind of test with ghci, first you MUST set `MonomorphismRestriction` and should use `:sprint`:

```haskell
% ghci 
> :set -XMonomorphismRestriction
> let a = (1+2,3+4)
> :sprint a
a = (_,_)
> :set -XStrict
> let b = (b1,3+4) where b1 = 1 + 2
> :sprint b
b = (3,_)
```

Reply:
> Or are you assuming that you're also enabling Strict by default???

Yes, in the cabal file.

Reply:
> But I see that the 7.x builds are generating spurious warnings from "cabal check", despite the "impl(ghc) >= 8" guard. Is there some way to avoid that, and have all the builds be green?

I guess it's time to drop the support for GHC 7.x.